text/0105-load-based-replica-read.md

```diff
@@ -0,0 +1,98 @@
+
+Replica read, which enables us to read from follower and learner peers, is a feature that can extend resources instantly. However, to preserve linearizability, we have to know the latest index of the leader before reading at a non-leader peer, making replica read not a good choice in all cases.
+
+Therefore, we hope to enable replica read intelligently based on the load of the leader. When all the read pools are vacant, all requests will still be handled by the leader for low latency. But if the read pool of some TiKV is too busy to handle requests in time, the client will automatically switch to other peers to utilize extra resources.
+
+### Estimating load of TiKV read pool
+
+Traditionally, the load is ratio of the average queue length and the processing speed. But we care more about the latency and latency is a metric that is more comparable across nodes. So, we will use the wait duration to represent the load.
+
+Knowing the current queue length $L$ and the average time slice S$, we can estimate that the wait duration is $T_{waiting} =L \cdot S$.
+
+The current queue length is easily known. But we have to predict the average time slice in the short future. We can use the EWMA of the previous time slices to estimate it:
+
+$$S_{i}=\alpha \cdot S_{now}+(1-\alpha) \cdot S_{i-1}$$
+
+### Rejecting request on busy
+
+The client decides the maximum waiting duration. If this field is set, TiKV can return `ServerIsBusy` early if the estimated waiting duration exceeds the threshold. This threshold is also configurable by the user.
+
+```protobuf
+message Context {
+    ...
+
+    uint32 busy_threshold_ms = 26;
+}
+```
+
+If TiKV returns `ServerIsBusy` based on the current load, the estimated waiting duration and the current index on the leader will be also returned to the client.
+
+```protobuf
+message ServerIsBusy {
+    ...
+
+    uint32 estimated_wait_ms = 3;
+    uint64 applied_index = 4;
+}
+```
+
+Because we will retry in replica-read mode, we don't need the follower or learner to issue a read index RPC again after knowing the applied index.
+
+The estimated waiting duration may be useful for the client to select a better TiKV for the retried request.
+
+### Retry strategy
+
+_It is hard to find a perfect strategy. It is expected that the strategy changes after further experiemnts. So, this part only describes a possible solution._
+
+Without any other information, the client will select a random follower or learner to retry the request after the client receives a `ServerIsBusy`. The `busy_threshold_ms` in the retried request will be set to _2 times_ the one returned in `ServerIsBusy`. So, if the retried TiKV is _much_ busier than the leader, it will reject the request again. The factor is to balance the extra RPC number and the load difference between TiKVs.
```

Comment:
Maybe round-robin is ok by now as the retry happens infrequently.

Reply:
Round-robin is okay if other instances are mostly idle. I worry that some followers are already busy. I agree the priority of balancing load between followers is relatively low.

However, I think it's necessary to let the follower reject when it's busier than the leader. In my demo, I didn't implement this, then I find the load on the followers are much much higher than the leader, because the leader load is limited by the threshold we set, but there is no limit on the follower.

Reply:
It would be better If the leader could decide if it would be worse to reject the current request more "accurately", but this looks difficult since the statistical information is not that up to date. 

How about we still maintain load statistic information for different nodes on the client side, it could help reduce the worse retry possibility to a certain extent though the information may not be quite accurate. If we want more up-to-date and accurate information maybe we also collect it from each node more actively or aggressively.

Reply:
The strategies according to load statistics are mainly described in the "Optimizing strategy with load info" section.

I just added a small optimization for the leader to reject fewer requests.